command/agent/command.go

```diff
@@ -547,13 +547,49 @@ func (c *Command) Run(args []string) int {
 	// Start retry join process
 	c.retryJoinErrCh = make(chan struct{})
 
-	joiner := retryJoiner{
-		join:     c.agent.server.Join,
-		discover: &discover.Discover{},
-		errCh:    c.retryJoinErrCh,
-		logger:   c.agent.logger,
+	if config.Server.Enabled && len(config.Server.RetryJoin) != 0 {
+		joiner := retryJoiner{
+			discover:      &discover.Discover{},
+			errCh:         c.retryJoinErrCh,
+			logger:        c.agent.logger,
+			serverJoin:    c.agent.server.Join,
+			serverEnabled: true,
+		}
+
+		// This is for backwards compatibility, this should be removed in Nomad
+		// 0.10 and only ServerJoin should be declared on the server
+		serverJoinInfo := &ServerJoin{
+			RetryJoin:        config.Server.RetryJoin,
+			StartJoin:        config.Server.StartJoin,
+			RetryMaxAttempts: config.Server.RetryMaxAttempts,
+			RetryInterval:    config.Server.RetryInterval,
+		}
+		go joiner.RetryJoin(serverJoinInfo)
+	}
+
```

Comment:
Seems like only one of the two (either config.Server.RetryJoin or config.Server.RetryJoin) should be set? If both are set it should fail validation

Reply:
Fixed in adding a `Validate` method for all edge cases like this.